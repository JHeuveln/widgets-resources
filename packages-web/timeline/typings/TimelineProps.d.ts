/**
 * This file was generated from Timeline.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { DynamicValue, EditableValue, ListValue, ObjectItem, WebImage } from "mendix";

export type RenderModeEnum = "basic" | "custom";

export interface TimelineContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex: number;
    renderMode: RenderModeEnum;
    data: ListValue;
    eventTime: (item: ObjectItem) => EditableValue<Date>;
    showDayDivider: boolean;
    title?: (item: ObjectItem) => EditableValue<string>;
    description?: (item: ObjectItem) => EditableValue<string>;
    icon?: DynamicValue<WebImage>;
    customIcon?: (item: ObjectItem) => ReactNode;
    customDayDivider?: (item: ObjectItem) => ReactNode;
    customTitle?: (item: ObjectItem) => ReactNode;
    customEventDateTime?: (item: ObjectItem) => ReactNode;
    customDescription?: (item: ObjectItem) => ReactNode;
}

export interface TimelinePreviewProps {
    class: string;
    style: string;
    renderMode: RenderModeEnum;
    data: {} | null;
    eventTime: string;
    showDayDivider: boolean;
    title: string;
    description: string;
    icon: string;
    customIcon: { widgetCount: number; renderer: ComponentType };
    customDayDivider: { widgetCount: number; renderer: ComponentType };
    customTitle: { widgetCount: number; renderer: ComponentType };
    customEventDateTime: { widgetCount: number; renderer: ComponentType };
    customDescription: { widgetCount: number; renderer: ComponentType };
}
