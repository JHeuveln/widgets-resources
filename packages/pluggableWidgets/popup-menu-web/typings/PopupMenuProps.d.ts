/**
 * This file was generated from PopupMenu.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { ActionValue } from "mendix";

export type TriggerEnum = "onclick" | "onhover";

export type RenderModeEnum = "basic" | "custom";

export type PositionEnum = "left" | "right" | "top" | "bottom";

export type ItemTypeEnum = "item" | "divider";

export type StyleClassEnum = "defaultStyle" | "primaryStyle" | "dangerStyle" | "customStyle";

export interface BasicItemsType {
    itemType: ItemTypeEnum;
    caption: string;
    action?: ActionValue;
    styleClass: StyleClassEnum;
}

export interface CustomItemsType {
    content: ReactNode;
    action?: ActionValue;
}

export interface BasicItemsPreviewType {
    itemType: ItemTypeEnum;
    caption: string;
    action: {} | null;
    styleClass: StyleClassEnum;
}

export interface CustomItemsPreviewType {
    content: { widgetCount: number; renderer: ComponentType };
    action: {} | null;
}

export interface PopupMenuContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex: number;
    menuTrigger?: ReactNode;
    trigger: TriggerEnum;
    renderMode: RenderModeEnum;
    position: PositionEnum;
    basicItems: BasicItemsType[];
    customItems: CustomItemsType[];
}

export interface PopupMenuPreviewProps {
    class: string;
    style: string;
    menuTrigger: { widgetCount: number; renderer: ComponentType };
    trigger: TriggerEnum;
    renderMode: RenderModeEnum;
    position: PositionEnum;
    basicItems: BasicItemsPreviewType[];
    customItems: CustomItemsPreviewType[];
}
