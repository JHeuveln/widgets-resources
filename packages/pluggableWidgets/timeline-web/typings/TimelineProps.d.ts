/**
 * This file was generated from Timeline.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties } from "react";
import { DynamicValue, ListValue, ListAttributeValue, ListWidgetValue, WebIcon } from "mendix";

export type RenderModeEnum = "basic" | "custom";

export type GroupByKeyEnum = "day" | "month" | "year";

export interface TimelineContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex: number;
    renderMode: RenderModeEnum;
    data: ListValue;
    eventTime: ListAttributeValue<Date>;
    showGroupDivider: boolean;
    groupByKey: GroupByKeyEnum;
    title?: ListAttributeValue<string>;
    description?: ListAttributeValue<string>;
    icon?: DynamicValue<WebIcon>;
    customIcon?: ListWidgetValue;
    customGroupDivider?: ListWidgetValue;
    customTitle?: ListWidgetValue;
    customEventDateTime?: ListWidgetValue;
    customDescription?: ListWidgetValue;
}

export interface TimelinePreviewProps {
    class: string;
    style: string;
    renderMode: RenderModeEnum;
    data: {} | null;
    eventTime: string;
    showGroupDivider: boolean;
    groupByKey: GroupByKeyEnum;
    title: string;
    description: string;
    icon: { type: "glyph"; iconClass: string } | { type: "image"; imageUrl: string } | null;
    customIcon: { widgetCount: number; renderer: ComponentType };
    customGroupDivider: { widgetCount: number; renderer: ComponentType };
    customTitle: { widgetCount: number; renderer: ComponentType };
    customEventDateTime: { widgetCount: number; renderer: ComponentType };
    customDescription: { widgetCount: number; renderer: ComponentType };
}
