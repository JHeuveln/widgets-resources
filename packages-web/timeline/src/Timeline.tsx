import { Component, ReactNode, createElement } from "react";
import { HelloWorldSample } from "./components/HelloWorldSample";

import { TimelineContainerProps } from "../typings/TimelineProps";

import "./ui/Timeline.css";

export default class Timeline extends Component<TimelineContainerProps> {
    render(): ReactNode {
        return <HelloWorldSample sampleText={this.props.sampleText ? this.props.sampleText : "World"} />;
    }
}
