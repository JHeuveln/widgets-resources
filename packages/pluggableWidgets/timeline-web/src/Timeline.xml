<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.timeline.Timeline" pluginWidget="true" needsEntityContext="true" offlineCapable="true"
        supportedPlatform="Web"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Timeline</name>
    <description>Shows vertical timeline with events</description>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="renderMode" type="enumeration" defaultValue="basic">
                    <caption>Render</caption>
                    <description>Basic: creates the Timeline based on the configured items. Custom: free to model timeline.</description>
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="data" type="datasource" isList="true">
                    <caption>Data Source</caption>
                    <description/>
                </property>
                <property key="eventTime" type="attribute" dataSource="data">
                    <caption>Event date time attribute</caption>
                    <description>Will be used for grouping by date.</description>
                    <attributeTypes>
                        <attributeType name="DateTime"/>
                    </attributeTypes>
                </property>
                <property key="showDayDivider" type="boolean" defaultValue="false">
                    <caption>Show day divider</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Event">
                <property key="title" type="attribute" dataSource="data" required="false">
                    <caption>Title</caption>
                    <description/>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
                <property key="description" type="attribute" dataSource="data" required="false">
                    <caption>Description</caption>
                    <description/>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
                <property key="icon" type="icon" required="false">
                    <caption>Icon</caption>
                    <description>Will be circle if its not selected.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Custom">
                <property key="customIcon" type="widgets" dataSource="data" required="false">
                    <caption>Icon</caption>
                    <description>Content of the icon</description>
                </property>
                <property key="customDayDivider" type="widgets" dataSource="data" required="false">
                    <caption>Day Divider</caption>
                    <description>Content of the day divider</description>
                </property>
                <property key="customTitle" type="widgets" dataSource="data" required="false">
                    <caption>Title</caption>
                    <description>Content of the title</description>
                </property>
                <property key="customEventDateTime" type="widgets" dataSource="data" required="false">
                    <caption>Event Time</caption>
                    <description>Content of the event time</description>
                </property>
                <property key="customDescription" type="widgets" dataSource="data" required="false">
                    <caption>Content</caption>
                    <description>Content of the description</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
