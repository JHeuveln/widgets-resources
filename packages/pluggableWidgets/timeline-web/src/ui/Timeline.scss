$primary-color: #264ae5;
$border-color: #dadcde;
$title-color: #0a1325;
$description-color: #6c717c;
$spacing-large: 30px;
$gutter-size: 15px;
$icon-size: 18px;
$icon-position-left: -($icon-size / 2);

.timeline-date {
    .timeline-date-header {
        display: inline-block;
        width: 110px;
        padding: 8px;
        text-align: center;
        border: 1px solid $border-color;
        border-radius: 30px;
    }
    .timeline-events {
        margin-bottom: 0;
        margin-left: 55px;
        padding: 35px 0;
        border-left: 1px solid $border-color;
        ul {
            padding: 0;
            list-style: none;
            margin-bottom: 0;
        }
        .timeline-event {
            position: relative;
            padding-left: ($gutter-size * 2);
            padding-bottom: 20px;

            .icon-wrapper {
                position: absolute;
                top: 3px;
                left: $icon-position-left;
                display: flex;
                justify-content: center;
                align-items: center;
                width: $icon-size;
                height: $icon-size;

                .glyphicon {
                    font-size: $icon-size;
                }
                img {
                    max-width: $icon-size;
                    max-height: $icon-size;
                }
            }

            .timeline-icon-circle {
                width: $icon-size;
                height: $icon-size;
                border-radius: 50%;
                background-color: $primary-color;
            }

            .content-wrapper {
                .date-time-wrapper {
                    margin-right: 15px;
                    p {
                    }
                }
                .info-wrapper {
                    flex-direction: column;

                    .title {
                        line-height: 20px;
                        color: $title-color;
                    }
                    .description {
                        line-height: 20px;
                        color: $description-color;
                    }
                }
            }
        }
    }

    .no-divider {
        padding-bottom: 0;
        padding-top: 0;
    }
}

.flex-container {
    display: flex;
    overflow: hidden;
    flex: 1;
    flex-direction: row;
}
