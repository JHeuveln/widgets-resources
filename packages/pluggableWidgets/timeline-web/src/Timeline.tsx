import { createElement, ReactElement, ReactNode, useMemo } from "react";
import { TimelineContainerProps } from "../typings/TimelineProps";
import "./ui/Timeline.scss";
import { DynamicValue, WebIcon } from "mendix";
import TimelineComponent from "./components/TimelineComponent";

export interface BasicItemType {
    icon?: DynamicValue<WebIcon>;
    title?: string;
    date?: string;
    time?: string;
    description?: string;
}

export interface CustomItemType {
    icon?: ReactNode;
    dayDivider?: ReactNode;
    title?: ReactNode;
    eventDateTime?: ReactNode;
    description?: ReactNode;
}

export type ItemType = BasicItemType | CustomItemType;

export default function Timeline(props: TimelineContainerProps): ReactElement {
    const structuredEvents = useMemo((): Map<string | ReactNode, ItemType[]> => {
        const eventsMap = new Map<string | ReactNode, ItemType[]>();

        props.data.items?.forEach(item => {
            let constructedItem: ItemType;
            const eventTime = props.eventTime(item);
            const date = eventTime.value;
            if (!date) {
                return;
            }
            let dateFormatter = eventTime.formatter;
            let timeFormatter = eventTime.formatter;

            if (dateFormatter?.type === "datetime") {
                dateFormatter = dateFormatter.withConfig({ type: "date" });
                timeFormatter = dateFormatter.withConfig({ type: "time" });
            }

            if (props.renderMode === "basic") {
                constructedItem = {
                    icon: props.icon,
                    title: props.title?.(item).displayValue,
                    date: dateFormatter?.format(eventTime.value),
                    time: timeFormatter?.format(eventTime.value),
                    description: props.description?.(item).displayValue
                };
            } else {
                constructedItem = {
                    icon: props.customIcon?.(item),
                    dayDivider: props.customDayDivider?.(item),
                    title: props.customTitle?.(item),
                    eventDateTime: props.customEventDateTime?.(item),
                    description: props.customDescription?.(item)
                };
            }

            const currentDates = eventsMap.get(dateFormatter?.format(eventTime.value));
            if (currentDates) {
                currentDates.push(constructedItem);
            } else {
                eventsMap.set(dateFormatter?.format(eventTime.value), [constructedItem]);
            }
        });

        return eventsMap;
    }, [props.data]);

    return (
        <TimelineComponent
            data={structuredEvents}
            showDayDivider={props.showDayDivider}
            renderMode={props.renderMode}
        />
    );
}
