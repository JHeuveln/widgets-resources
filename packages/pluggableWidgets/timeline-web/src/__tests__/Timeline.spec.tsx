import { shallow } from "enzyme";
import { createElement } from "react";
import { TimelineContainerProps } from "../../typings/TimelineProps";
import { EditableValueBuilder, ListValueBuilder } from "@widgets-resources/piw-utils";
import Timeline from "../Timeline";

jest.mock("mendix/components/web/Icon", () => jest.requireActual("./__mocks__/mendix/components/web/Icon"));

describe("Timeline", () => {
    const listValueBuilder = ListValueBuilder();
    const defaultProps: TimelineContainerProps = {
        name: "timeline",
        class: "test",
        tabIndex: 0,
        renderMode: "basic",
        data: listValueBuilder.simple(),
        eventTime: () => new EditableValueBuilder<Date>().withValue(new Date(1453, 4, 29)).build(),
        showDayDivider: true,
        title: () => new EditableValueBuilder<string>().withValue("title").build(),
        description: () => new EditableValueBuilder<string>().withValue("description").build(),
        customIcon: () => <img src={"test"} />,
        customDayDivider: () => <p>Custom Divider</p>,
        customTitle: () => <p>Custom Title</p>,
        customEventDateTime: () => <p>Custom Event Date Time</p>,
        customDescription: () => <p>Custom Description</p>
    };
    it("renders correctly with default props", () => {
        const component = shallow(<Timeline {...defaultProps} />);
        expect(component).toMatchSnapshot();
    });
});
