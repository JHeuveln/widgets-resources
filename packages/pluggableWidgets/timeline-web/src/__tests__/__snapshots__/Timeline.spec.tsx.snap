// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline converts the data in expected map 1`] = `
<TimelineComponent
  data={
    Map {
      "Formatted Sun May 29 1453 00:00:00 GMT+0019 (Central European Summer Time)" => Array [
        Object {
          "description": "Formatted description",
          "icon": undefined,
          "time": "Formatted Sun May 29 1453 00:00:00 GMT+0019 (Central European Summer Time)",
          "title": "Formatted title",
        },
        Object {
          "description": "Formatted description",
          "icon": undefined,
          "time": "Formatted Sun May 29 1453 00:00:00 GMT+0019 (Central European Summer Time)",
          "title": "Formatted title",
        },
      ],
    }
  }
  renderMode="basic"
  showDayDivider={true}
/>
`;
